source-mongodb/pull.go

```diff
@@ -132,141 +109,213 @@ func (d *driver) Pull(open *pc.Request_Open, stream *boilerplate.PullOutput) err
-	BackfillLastId bson.RawValue `json:"backfill_last_id,omitempty"`
+	StartedAt time.Time `json:"started_at,omitempty"`
+	LastId bson.RawValue `json:"last_id,omitempty"`
+}
 
-	StreamResumeToken bson.Raw `json:"stream_resume_token,omitempty"`
+type resourceState struct {
+	Backfill backfillState `json:"backfill,omitempty"`
 }
 
 type docKey struct {
 	Id interface{} `bson:"_id"`
 }
+type namespace struct {
+	Database string `bson:"db"`
+	Collection string `bson:"coll"`
+}
 type changeEvent struct {
-	DocumentKey   docKey `bson:"documentKey"`
-	OperationType string `bson:"operationType"`
-	FullDocument  bson.M `bson:"fullDocument"`
+	DocumentKey   docKey    `bson:"documentKey"`
+	OperationType string    `bson:"operationType"`
+	FullDocument  bson.M    `bson:"fullDocument"`
+	Ns            namespace `bson:"ns"`
 }
 
 const resumePointGoneErrorMessage = "the resume point may no longer be in the oplog"
 const resumeTokenNotFoundErrorMessage = "the resume token was not found"
+const unauthorizedMessage = "not authorized on admin to execute command"
 
-func (c *capture) ChangeStream(ctx context.Context, client *mongo.Client, binding int, res resource, state *resourceState) error {
-	var db = client.Database(res.Database)
+func (c *capture) ChangeStream(ctx context.Context, client *mongo.Client, resources []resource, state captureState) error {
+	if len(resources) < 1 {
+		return nil
+	}
 
-	var collection = db.Collection(res.Collection)
+	// If we have a stream resume token, we use that to continue our change stream
+	// otherwise, we look at Backfill.StartedAt of all resources and use the
+	// minimum value to start our change stream from
+	var streamStartAt time.Time
+	if len(state.StreamResumeToken) < 1 {
+		for _, res := range resources {
+			var resState = state.Resources[resourceId(res)]
+
+			if !resState.Backfill.StartedAt.IsZero() {
+				if streamStartAt.IsZero() {
+					streamStartAt = resState.Backfill.StartedAt
+				} else if resState.Backfill.StartedAt.Before(streamStartAt) {
+					streamStartAt = resState.Backfill.StartedAt
+				}
+			}
+		}
+	}
+
+	var collectionBindingIndex = make(map[string]int)
+	for idx, res := range resources {
+		collectionBindingIndex[resourceId(res)] = idx
+	}
 
 	log.WithFields(log.Fields{
-		"database":           res.Database,
-		"collection":         res.Collection,
-		"state":              state,
-	}).Info("listening on changes on collection")
-	var eventFilter = bson.D{{"$match", bson.D{{"$or",
-	bson.A{
-		bson.D{{"operationType", "delete"}},
-		bson.D{{"operationType", "insert"}},
-		bson.D{{"operationType", "update"}},
-		bson.D{{"operationType", "replace"}},
-	}}}}}
+		"state":             state,
+		"streamResumeToken": state.StreamResumeToken,
+		"streamStartAt":     streamStartAt,
+	}).Info("listening on changes")
+
+	var eventFilter = bson.D{{"$match", bson.D{{
+		"$and", bson.A{
+			bson.D{{"$or",
+				bson.A{
+					bson.D{{"operationType", "delete"}},
```

Comment:
At the risk of being pedantic...are there _other_ events we should be listening for? Should we filter out any events? It seems like it would be best to take everything the database will give us and do all of the filtering in the connector, extracting a resume token even from non CRUD events.